### ISSUE TITLE
Refactor YulUtilFunctions::revertReasonIfDebug to return an actual function

### ISSUE THREAD

**Issue by chriseth** [ethereum/solidity#8456]
YulUtilFunctions::revertReasonIfDebug currently returns code while all other functions in that class return a function. This can be refactored now that we have `CompilerContext::callYulFunction`

**Comment by mijovic** [comment]
A lot of times `YulUtilFunctions::revertReasonIfDebug` is called from `Whiskers` in inline assembly blocks, so it is not really possible to call with `CompilerContext::callYulFunction`
Maybe we can write similar function in `CompilerContext` that calls `YulUtilFunction`s from inline assembly as well

**Comment by chriseth** [comment]
Maybe we should also just keep it inlined in the old code generator (is that what you meant?).

**Comment by mijovic** [comment]
> Maybe we should also just keep it inlined in the old code generator (is that what you meant?).

Sample usage, for example in `CompilerUtils.cpp` around like 291
```
Whiskers templ(R"({
	if gt(ptr, 0x100000000) { <revertStringPointer> }
	...
	if gt(array_data_start, input_end) { <revertStringStart> }
	...
	mstore(dst, array_length)
	dst := add(dst, 0x20)
})");
templ("revertStringPointer", m_context.revertReasonIfDebug("ABI memory decoding: invalid data pointer"));
...
m_context.appendInlineAssembly(templ.render(), {"input_end", "base_offset", "offset", "ptr", "dst"});
```

So, maybe to inline it via `revertReasonIfDebug` (as a return string) in CompilerContext, and using function from `YulUtilFunctions` where possible. Bad thing is that if we do like this we will have same code on two places

**Comment by axic** [comment]
@mijovic since you worked on #10765, how much work do you think this issue would be?

**Comment by mijovic** [comment]
I tried to fix this with #9195 some time ago, but the actual problem is that most of the time this is called from inline assembly for old codegen.
And I think it makes sense to do this after #8559 or something similar. The think is that inline assembly doesn't see yul util functions...

However for IR this would be easy I think
